service/storage_service.cc

```diff
@@ -6176,6 +6176,162 @@ static bool increases_replicas_per_rack(const locator::topology& topology, const
+
+    auto start = std::chrono::steady_clock::now();
+    slogger.info("Starting tablet repair by API request table_id={} tokens={} all_tokens={} tablet_task_id={}", table, tokens, all_tokens, tablet_task_info.tablet_task_id);
+
+    while (true) {
+        auto guard = co_await wait_for_tablet_migration();
+
+        auto& tmap = get_token_metadata().tablets().get_tablet_map(table);
+        std::vector<canonical_mutation> updates;
+
+        if (all_tokens) {
+            tokens.clear();
+            co_await tmap.for_each_tablet([&] (locator::tablet_id tid, const locator::tablet_info& info) -> future<> {
+                auto last_token = tmap.get_last_token(tid);
+                tokens.push_back(last_token);
+                co_return;
+            });
+        }
+
+        for (const auto& token : tokens) {
+            auto tid = tmap.get_tablet_id(token);
+            if (tmap.get_tablet_transition_info(tid)) {
+                throw std::runtime_error(fmt::format("Tablet {} is in transition", locator::global_tablet_id{table, tid}));
+            }
+            auto& tinfo = tmap.get_tablet_info(tid);
+            auto& req_id = tinfo.tablet_task_info.tablet_task_id;
+            if (req_id) {
+                throw std::runtime_error(fmt::format("Tablet {} is already in repair by tablet_task_id={}",
+                        locator::global_tablet_id{table, tid}, req_id));
+            }
+            auto last_token = tmap.get_last_token(tid);
+            updates.emplace_back(
+                replica::tablet_mutation_builder(guard.write_timestamp(), table)
+                    .set_tablet_task_info(last_token, tablet_task_info)
+                    .build());
+        }
+
+        sstring reason = format("Repair tablet by API request tokens={} tablet_task_id={}", tokens, tablet_task_info.tablet_task_id);
+        if (co_await exec_tablet_migration(std::move(guard), std::move(updates), std::move(reason))) {
+            break;
+        }
+    }
+
+    co_await _topology_state_machine.event.wait([&] {
+        auto& tmap = get_token_metadata().tablets().get_tablet_map(table);
+        return std::all_of(tokens.begin(), tokens.end(), [&] (const dht::token& token) {
+            auto id = tmap.get_tablet_id(token);
+            return tmap.get_tablet_info(id).tablet_task_info.tablet_task_id != tablet_task_info.tablet_task_id;
+        });
+    });
+
+    auto duration = std::chrono::duration<float>(std::chrono::steady_clock::now() - start);
+    slogger.info("Finished tablet repair by API request table_id={} tokens={} all_tokens={} tablet_task_id={} duration={}",
+            table, tokens, all_tokens, tablet_task_info.tablet_task_id, duration);
+
+    co_return res;
+}
+
+
+// Delete a tablet repair request by the given tablet_task_id
+future<> storage_service::del_repair_tablet_request(table_id table, locator::tablet_task_id tablet_task_id) {
```

Comment:
It is intended to remove a previous repair request by user. We need a way to modify the tmap from any node in the cluster.  If we have better ways  other than  sending mutations over "group0->client().add_entry", I'd be happy to use.

Reply:
group0 update is fine, my concern is that if this is to cancel things, then it's not effective, because it will wait for the whole repair to finish before it cancels:

> For cancellation, we have to delete tasks and repair transitions without waiting for migration to finish.

Reply:
Currently, we wait for the state machine to be in tablet_migration in 1. 
```
storage_service::del_repair_tablet_request
      1 wait_for_tablet_migration
      2 create updates to delete the del_tablet_task_info
      3 add_entry
```

To delete the tablet task info, can we skip 1 and run 2 and 3 directly? I think we can.

Reply:
I think we should avoid deleting tablet transition outside the regular state machine handling code. It is asking for trouble to do it in multiple places. I will handle it in the repair transition handler in case the task is removed by user.